### PR TITLE
feat(perf_counter): add a perf_counter of server session

### PR DESCRIPTION
**Pull request by Smityz**
Add a perf counter to monitor the session kept on the server.
```
[server*network*client_session_count, NUMBER, 3.00]
[server*network*client_session_count, NUMBER, 4.00]
[server*network*client_session_count, NUMBER, 4.00]
[server*network*client_session_count, NUMBER, 4.00]
[server*network*client_session_count, NUMBER, 5.00]
[server*network*client_session_count, NUMBER, 14.00]
[server*network*client_session_count, NUMBER, 14.00]
[server*network*client_session_count, NUMBER, 4.00]
```

### CODE REVIEW

File: include/dsn/tool-api/network.h [hunk 1]
```diff
@@ -40,6 +41,8 @@ class rpc_engine;
 class service_node;
 class task_worker_pool;
 class task_queue;
+class perf_counter_wrapper;
```
Comment (levy5307) [inline include/dsn/tool-api/network.h:44]: I think this line is useless

File: src/runtime/rpc/network.cpp [hunk 2]
```diff
@@ -590,6 +590,11 @@ connection_oriented_network::connection_oriented_network(rpc_engine *srv, networ
     : network(srv, inner_provider)
 {
     _cfg_conn_threshold_per_ip = 0;
+    _client_session_count.init_global_counter("replica",
+                                              "network",
+                                              "server_session_count",
+                                              COUNTER_TYPE_NUMBER,
+                                              "current session count on server");
```
Comment (hycdong) [inline src/runtime/rpc/network.cpp:597]: I think `server.network.client_session_count` maybe better. 
Firstly, both replica server and meta server will share this counter, `server` maybe more suitable.  Besides, `client_session_count` or `session_count` is better.

File: src/runtime/rpc/network.cpp [hunk 3]
```diff
@@ -663,8 +669,8 @@ rpc_session_ptr connection_oriented_network::get_server_session(::dsn::rpc_addre
 
 void connection_oriented_network::on_server_session_accepted(rpc_session_ptr &s)
 {
-    int scount = 0;
-    int ecount = 1;
+    int ip_count = 0;
+    int ip_conn_count = 1;
```
Comment (hycdong) [inline src/runtime/rpc/network.cpp:673]: Could you please adding some comments to explain the difference between `ip_count` and `ip_conn_count`?

**Comment by hycdong** [comment]
In my view, to record the server session count, when came a new client session,  `_client_session_count++`, when a client disconnect, `_client_session_count++`. However, I only saw `_client_session_count->set(ip_count);` in this patch to set the counter value. Could you please explain how to record the right session count?

File: src/runtime/rpc/network.cpp (hunk 3 above)
Comment (Smityz) [reply src/runtime/rpc/network.cpp:673]: https://github.com/XiaoMi/rdsn/blob/a1e831b6a2c8cefe7e24eb5259d146818d4d2ef9/src/runtime/rpc/network.cpp#L708
The description of these names already existed in this file

**Comment by Smityz** [comment]
> In my view, to record the server session count, when came a new client session, `_client_session_count++`, when a client disconnect, `_client_session_count++`. However, I only saw `_client_session_count->set(ip_count);` in this patch to set the counter value. Could you please explain how to record the right session count?

I just add a perf counter below the original log, its correctness is guaranteed by the log.

**Review by Shuo-Jia — APPROVED** [review]

File: src/runtime/rpc/network.cpp (hunk 3 above)
Comment (hycdong) [reply src/runtime/rpc/network.cpp:673]: Well, the meaning of `ip_count` is clear. I just don't know the meaning of `ip_conn_count`, especially with the comment below.
https://github.com/XiaoMi/rdsn/blob/a1e831b6a2c8cefe7e24eb5259d146818d4d2ef9/src/runtime/rpc/network.cpp#L709

**Comment by hycdong** [comment]
> > In my view, to record the server session count, when came a new client session, `_client_session_count++`, when a client disconnect, `_client_session_count++`. However, I only saw `_client_session_count->set(ip_count);` in this patch to set the counter value. Could you please explain how to record the right session count?
> 
> I just add a perf counter below the original log, its correctness is guaranteed by the log.

I just don't figure out the original code logic :-( , and I wonder if you can explain it to me ^_^

**Comment by Smityz** [comment]
> > > In my view, to record the server session count, when came a new client session, `_client_session_count++`, when a client disconnect, `_client_session_count++`. However, I only saw `_client_session_count->set(ip_count);` in this patch to set the counter value. Could you please explain how to record the right session count?
> > 
> > 
> > I just add a perf counter below the original log, its correctness is guaranteed by the log.
> 
> I just don't figure out the original code logic, :-( , and I wonder if you can explain it to me ^_^.

I add more comments here https://github.com/XiaoMi/rdsn/pull/894/files#diff-8e070378de2a59fce981099050b87a61f246697846a65c568ca5b3a2798d3bedL700
Hope that can help you

**Comment by Smityz** [comment]
> In my view, to record the server session count, when came a new client session, `_client_session_count++`, when a client disconnect, `_client_session_count++`. However, I only saw `_client_session_count->set(ip_count);` in this patch to set the counter value. Could you please explain how to record the right session count?

add count:
https://github.com/XiaoMi/rdsn/blob/fcd899333a0f3eee526d32d58e3cff1f48b1baab/src/runtime/rpc/network.cpp#L677
minus count:
https://github.com/XiaoMi/rdsn/blob/fcd899333a0f3eee526d32d58e3cff1f48b1baab/src/runtime/rpc/network.cpp#L719

all sessions are kept in a map, so we can get the total count according to the map size

**Review by hycdong — APPROVED** [review]

**Review by Shuo-Jia — APPROVED** [review]

**Pull request merged**